internal/lib/lib.go

```diff
@@ -980,3 +981,107 @@ func IsAviLBDefaultIngressClassWithClient(kc kubernetes.Interface) (string, bool
 	utils.AviLog.Debugf("IngressClass with controller ako.vmware.com/avi-lb not found in the cluster")
 	return "", false
 }
+
+func GetControllerPropertiesFromSecret(kc kubernetes.Interface) (ctrlUsername, ctrlPassword, ctrlAuthToken string, err error) {
+	aviSecret, err := kc.CoreV1().Secrets("avi-system").Get(context.TODO(), "avi-secret", metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	ctrlUsername = string(aviSecret.Data["username"])
+	if aviSecret.Data["password"] != nil {
+		ctrlPassword = string(aviSecret.Data["password"])
+		return
+	}
+	if aviSecret.Data["authtoken"] != nil {
+		ctrlAuthToken = string(aviSecret.Data["authtoken"])
+		return
+	}
+	return
+}
+
+func GetAviSecretWithRetry(kc kubernetes.Interface, retryCount int) (*v1.Secret, error) {
+	var aviSecret *v1.Secret
+	var err error
+	for retry := 0; retry < retryCount; retry++ {
+		aviSecret, err = kc.CoreV1().Secrets(AviNS).Get(context.TODO(), AviSecret, metav1.GetOptions{})
+		if err == nil {
+			return aviSecret, nil
+		}
+		utils.AviLog.Warnf("Failed to get avi-secret, retry count:%d, err: %+v", retry, err)
+	}
+	return nil, err
+}
+
+func UpdateAviSecretWithRetry(kc kubernetes.Interface, aviSecret *v1.Secret, retryCount int) error {
+	var err error
+	for retry := 0; retry < retryCount; retry++ {
+		_, err = kc.CoreV1().Secrets(AviNS).Update(context.TODO(), aviSecret, metav1.UpdateOptions{})
+		if err == nil {
+			return nil
+		}
+		utils.AviLog.Warnf("Failed to update avi-secret, retry count:%d, err: %+v", retry, err)
+	}
+	return err
+}
+
+func RefreshAuthToken(kc kubernetes.Interface) {
+
+	retryCount := 5
+	ctrlIpAddress := os.Getenv("CTRL_IPADDRESS")
+	ctrlUsername, _, ctrlAuthToken, err := GetControllerPropertiesFromSecret(kc)
+	if err != nil {
+		utils.AviLog.Errorf("Failed to read from avi secret, err:", err)
```

Comment:
Shouldn't we return these errors from this function and handle this in the caller function called during bootup? If we get an error during bootup while refreshing the auth token, we should ideally handle this and restart AKO.

Reply:
the function call is now done only during boot up, and shutdown API is called if the secret read fails